src/self-managed-debug/src/system_catalog_dumper.rs

```diff
@@ -615,57 +615,231 @@ const RELATIONS: &[Relation] = &[
     },
 ];
 
-const TIMEOUT: Duration = Duration::from_secs(60);
+static PG_CONNECTION_TIMEOUT: Duration = Duration::from_secs(60);
+/// Timeout for a query. We use 6 minutes since it's a good
+/// sign that the operation won't work.
+static PG_QUERY_TIMEOUT: Duration = Duration::from_secs(20);
 
-#[derive(Debug, Clone)]
+/// The maximum number of errors we tolerate for a cluster replica.
+/// If a cluster replica has more than this many errors, we skip it.
+static MAX_CLUSTER_REPLICA_ERROR_COUNT: usize = 3;
+
+static SET_SEARCH_PATH_QUERY: &str = "SET search_path = mz_internal, mz_catalog, mz_introspection";
+static SELECT_CLUSTER_REPLICAS_QUERY: &str = "SELECT c.name as cluster_name, cr.name as replica_name FROM mz_clusters AS c JOIN mz_cluster_replicas AS cr ON c.id = cr.cluster_id;";
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct ClusterReplica {
-    // TODO (debug_tool1): Use this when scraping introspection relations
-    pub _cluster_name: String,
-    pub _replica_name: String,
+    pub cluster_name: String,
+    pub replica_name: String,
+}
+
+impl Default for ClusterReplica {
+    fn default() -> Self {
+        Self {
+            cluster_name: "mz_catalog_server".to_string(),
+            replica_name: "r1".to_string(),
+        }
+    }
+}
+
+impl fmt::Display for ClusterReplica {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}.{}", self.cluster_name, self.replica_name)
+    }
 }
 
 pub struct SystemCatalogDumper<'n> {
     context: &'n Context,
-    pg_client: Arc<PgClient>,
-    // TODO (debug_tool1): Use this when scraping introspection relations
-    _cluster_replicas: Vec<ClusterReplica>,
+    pg_client: Arc<Mutex<PgClient>>,
+    pg_tls: MakeTlsConnector,
+    cluster_replicas: Vec<ClusterReplica>,
     _pg_conn_handle: JoinHandle<Result<(), tokio_postgres::Error>>,
 }
 
+pub async fn create_postgres_connection(
+    connection_string: &str,
+) -> Result<
+    (
+        PgClient,
+        Connection<Socket, TlsStream<Socket>>,
+        MakeTlsConnector,
+    ),
+    anyhow::Error,
+> {
+    let mut pg_config = PgConfig::from_str(connection_string)?;
+    pg_config.connect_timeout(PG_CONNECTION_TIMEOUT);
```

Comment:
It might be nice to have the timeout be configurable via a CLI arg? Doesn't need to be fixed in this PR though!